intellij/src/saros/intellij/eventhandler/project/ProjectClosedHandler.java

```diff
@@ -0,0 +1,81 @@
+package saros.intellij.eventhandler.project;
+
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.project.ProjectManagerListener;
+import com.intellij.util.messages.MessageBusConnection;
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+import saros.repackaged.picocontainer.Disposable;
+import saros.session.ISarosSession;
+import saros.session.ISarosSessionManager;
+import saros.session.SessionEndReason;
+import saros.util.ThreadUtils;
+
+/**
+ * Ends the current session when a project containing shared modules is closed. This is done to
+ * avoid calls on disposed project objects.
+ *
+ * <p>The handler is enabled by default after instantiation.
+ */
+public class ProjectClosedHandler implements Disposable {
+
+  private static final Logger log = Logger.getLogger(ProjectClosedHandler.class);
+
+  private final ISarosSessionManager sarosSessionManager;
+  private final ISarosSession sarosSession;
+
+  private final MessageBusConnection messageBusConnection;
+
+  @SuppressWarnings("FieldCanBeLocal")
+  private final ProjectManagerListener projectManagerListener =
+      new ProjectManagerListener() {
+        @Override
+        public void projectClosing(@NotNull Project project) {
```

Comment:
Which thread calls this method ?

Reply:
This is run on the swing dispatcher thread: It is triggered by the [ProjectManagerImpl](https://github.com/JetBrains/intellij-community/blob/4fbb6152ee1391b117b85b464107bb87de23f6ab/platform/platform-impl/src/com/intellij/openapi/project/impl/ProjectManagerImpl.java#L701), which [asserts that the current thread is the dispatcher thread](https://github.com/JetBrains/intellij-community/blob/4fbb6152ee1391b117b85b464107bb87de23f6ab/platform/platform-impl/src/com/intellij/openapi/project/impl/ProjectManagerImpl.java#L665). (I also double-checked this by adding the same assert statement to the listener method.)

But just checking the *current* implementation is somewhat risky as I couldn't find anything in the documentation guaranteeing that this will always be the case, meaning they could change this internal behavior at some point.

Maybe we should add such assertions to sections of our code as well if we rely on the execution being done on the dispatcher thread and the execution itself is not triggered by us directly. This would make it much easier to detect changes to the internal IntelliJ logic that would break parts of our logic.

Reply:
The problem is that stopSession should not be called on the dispatch thread. It might be possible that are activities are executed in a dead session. The shutdown is still problematic anyways as we do not know then the PicoContainer calls stop on the ActivityHandler.

Reply:
Hmm ok, but we somehow have to deal with the project being closed. Even if there still are activities pending for a dead session, with the project closed, I don't think we have any way of applying them. (And even if there were a way of interacting with a closed module / a module of a closed project, I am not sure if that is something that we really want to use.)

I changed this to be a synchronous call as I was running into the mentioned problem with it being synchronous: The project object was already disposed (meaning the project was closed and torn down) but the session context was still alive and there were still pending activities. This caused exceptions as the pending activities were trying to use the disposed project object (as it is used for the hashcode/equals implementation of the module representation). This could be 'avoided' by saving the used values in the `IntelliJProjectImpl` directly instead of requesting them from the project, but the base of the problem (applying activities to closed modules) is still present. Furthermore, this causes problems when the used values are changed during runtime as the saved values would not be updated accordingly.

This change "solved" the issue in my trivial tests, but that is no indication that there isn't still the possibility of a race condition here.

The problem is that currently, I am not really able to test this well as Saros can't handle multiple open projects, meaning I can't test the long-term effects of the current implementation as the plugin is always torn down shortly after this is called (since the IntelliJ instance stopped). (Furthermore, doing manual tests for concurrency issues alone on one computer is somewhat complicated :wink:.)

Reply:
It is simply a design decision to start / stop the session from a non UI thread. Many components are coded with this contract. I know your intention to stop the session in this listener. As I do not have deep knowledge on this IntelliJ stuff there could be two approaches.

1. Shutdown the session in another thread and run the AWT/IntelliJ event loop in the meantime (I do not know if you are allowed to do that, i.e accessing the AWT EventQueue directly or IntelliJ offers such functionality). I also do not know how this work with different modal contexts. 

The other option would be to implement a IntelliJ callback for the ActivityHandler which discards activities for disposed projects and then shutdown the session in the background while you are performing a project switch.

Reply:
I don't think using the first approach is an acceptable solution as blocking the AWT event dispatcher thread blocks the entire IDE.

I considered the second option, but this just moves the problem: At least with the current implementation, filtering for modules of disposed project requires us to compare IProject objects, which in turn tries to call the IntelliJ project of the held module (which is what we are trying to avoid in the first place).
Furthermore, the session uses `hashCode` on the contained `SPath` when receiving/parsing activities, which also leads to exception as it subsequently calls `hashCode` on the `IProject` object.

It seem like I'l have to adjust the implantation anyway to somehow make it (IntelliJ) project object independent, at which point filtering activities in the session should (probably) work.